src/razor/src/dynamicFile/dynamicFileInfoHandler.ts

```diff
@@ -33,39 +37,57 @@ export class DynamicFileInfoHandler {
                 await this.removeDynamicFileInfo(request);
             }
         );
+        this.documentManager.onChange(async (e) => {
+            if (e.kind == RazorDocumentChangeKind.csharpChanged && !e.document.isOpen) {
+                const uriString = UriConverter.serialize(e.document.uri);
+                const identifier = TextDocumentIdentifier.create(uriString);
+                await vscode.commands.executeCommand(
+                    DynamicFileInfoHandler.dynamicFileUpdatedCommand,
+                    new RazorDynamicFileChangedParams(identifier)
+                );
+            }
+        });
     }
 
     // Given Razor document URIs, returns associated generated doc URIs
     private async provideDynamicFileInfo(
         request: ProvideDynamicFileParams
     ): Promise<ProvideDynamicFileResponse | null> {
-        let virtualUri: DocumentUri | null = null;
+        this.documentManager.roslynActivated = true;
+        const vscodeUri = vscode.Uri.parse(request.razorDocument.uri, true);
+
+        // Normally we start receiving dynamic info after Razor is initialized, but if the user had a .razor file open
+        // when they started VS Code, the order is the other way around. This no-ops if Razor is already initialized.
+        await this.documentManager.ensureRazorInitialized();
+
+        const razorDocument = await this.documentManager.getDocument(vscodeUri);
         try {
-            const vscodeUri = vscode.Uri.parse(request.razorDocument.uri, true);
-            const razorDocument = await this.documentManager.getDocument(vscodeUri);
             if (razorDocument === undefined) {
                 this.logger.logWarning(
                     `Could not find Razor document ${vscodeUri.fsPath}; adding null as a placeholder in URI array.`
                 );
-            } else {
-                // Retrieve generated doc URIs for each Razor URI we are given
-                const virtualCsharpUri = UriConverter.serialize(razorDocument.csharpDocument.uri);
-                virtualUri = virtualCsharpUri;
+
+                return null;
             }
 
-            this.documentManager.roslynActivated = true;
+            const virtualCsharpUri = UriConverter.serialize(razorDocument.csharpDocument.uri);
 
-            // Normally we start receiving dynamic info after Razor is initialized, but if the user had a .razor file open
-            // when they started VS Code, the order is the other way around. This no-ops if Razor is already initialized.
-            await this.documentManager.ensureRazorInitialized();
+            if (this.documentManager.isRazorDocumentOpenInCSharpWorkspace(vscodeUri)) {
+                // Open documents have didOpen/didChange to update the csharp buffer. Razor
+                // does not send edits and instead lets vscode handle them.
+                return new ProvideDynamicFileResponse({ uri: virtualCsharpUri }, null);
```

Comment:
Should this reset any stored edits, so that when a document is closed, things have a clean slate?

ie, when a document is opened, LSP will send a didOpen with the full contents, then any future edits, and then when its closed, are we at risk of sending some edits that were made before the open?